src/requestforattestation/RequestForAttestation.ts

```diff
@@ -99,39 +158,110 @@ export default class RequestForAttestation implements IRequestForAttestation {
   ) {
-    if (claim.owner !== identity.address) {
-      throw Error('Claim owner is not identity')
+    if (identity) {
+      if (requestInput.claim.owner !== identity.address) {
+        throw Error('Claim owner is not identity')
+      }
+      if (
+        !requestInput.claim ||
+        !requestInput.claimOwner ||
+        !requestInput.cTypeHash ||
+        !requestInput.claimHashTree
+      ) {
+        throw new Error(
+          `Property Not Provided while building RequestForAttestation with supplied Identity:\n
+          requestInput.claim:\n
+          ${requestInput.claim}\n
+          requestInput.claimOwner:\n
+          ${requestInput.claimOwner}\n
+          requestInput.cTypeHash:\n
+          ${requestInput.cTypeHash}\n
+          requestInput.claimHashTree:\n
+          ${requestInput.claimHashTree}`
+        )
+      }
+      this.legitimations = []
+      this.claim = requestInput.claim
+      this.claimOwner = requestInput.claimOwner
+      this.cTypeHash = requestInput.cTypeHash
+      if (
+        typeof requestInput.legitimations !== 'undefined' &&
+        Array.isArray(requestInput.legitimations) &&
+        requestInput.legitimations.length
+      ) {
+        requestInput.legitimations.forEach(element => {
+          const ele = AttestedClaim.fromAttestedClaim(element)
+          if (ele instanceof AttestedClaim) {
+            this.legitimations.push(ele)
+          }
+        })
+      }
+      this.delegationId = requestInput.delegationId
+      this.claimHashTree = requestInput.claimHashTree
+      this.rootHash = this.calculateRootHash()
+      this.claimerSignature = this.sign(identity)
+    } else if (requestInput.claimerSignature) {
+      if (
+        !requestInput.claim ||
+        !requestInput.claimOwner ||
+        !requestInput.claimHashTree ||
+        !requestInput.cTypeHash
+      ) {
+        throw new Error(
+          `Property Not Provided while building RequestForAttestation without Identity:\n
+          requestInput.claim:\n
+          ${requestInput.claim}\n
+          requestInput.claimOwner:\n
+          ${requestInput.claimOwner}\n
+          requestInput.cTypeHash:\n'
+          ${requestInput.cTypeHash}\n
+          requestInput.claimHashTree:\n
+          ${requestInput.claimHashTree}\n
+          requestInput.claimerSignature:\n
+          ${requestInput.claimerSignature}`
+        )
+      }
+      this.legitimations = []
+      this.claim = requestInput.claim
+      this.claimOwner = requestInput.claimOwner
+      this.cTypeHash = requestInput.cTypeHash
+      if (
+        typeof requestInput.legitimations !== 'undefined' &&
+        Array.isArray(requestInput.legitimations) &&
+        requestInput.legitimations.length
+      ) {
+        requestInput.legitimations.forEach(element => {
+          const ele = AttestedClaim.fromAttestedClaim(element)
+          if (ele instanceof AttestedClaim) {
+            this.legitimations.push(ele)
+          }
+        })
+      }
```

Comment:
Avoid using mutative methods such as `push`. Please feel free to ping me live for more details on that, if you're not sure why it's good practice / what I mean :) 

Instead you could do (safer and shorter, arguably clearer):

```javascript 
this.legitimations = requestInput.legitimations
                                    .map(l => AttestedClaim.fromAttestedClaim(l))
                                    .filter(c => c instanceof AttestedClaim);
```